File: go/transaction.go
package omne

import (
	"crypto/sha256"
	"fmt"
	"math/big"
)

const (
	DefaultGasLimit = uint64(200_000)
	DefaultGasPrice = "5000"
)

// Transaction is an unsigned Omne transaction. Value is a u128 (nil == 0);
// GasPrice is a decimal string; Data is hex-encoded ABI calldata.
type Transaction struct {
	From     string
	To       string
	Value    *big.Int
	GasLimit uint64
	GasPrice string
	Nonce    uint64
	ChainID  int
	Data     string
}

// SignedTransaction is a Transaction plus its ML-DSA-44 signature + public key.
type SignedTransaction struct {
	Transaction
	Signature string // hex (2420-byte sig)
	PublicKey string // hex (1312-byte pubkey)
}

func leUint(v uint64, n int) []byte {
	b := make([]byte, n)
	for i := 0; i < n; i++ {
		b[i] = byte(v & 0xff)
		v >>= 8
	}
	return b
}

func leBig(v *big.Int, n int) []byte {
	be := make([]byte, n)
	if v != nil {
		v.FillBytes(be) // big-endian, left-padded
	}
	le := make([]byte, n)
	for i := 0; i < n; i++ {
		le[i] = be[n-1-i]
	}
	return le
}

// HashTransaction computes the canonical 32-byte signing preimage hash,
// matching the Rust-side hash_transaction() and the TS/Python SDKs:
//
//	SHA-256( from(32) ‖ to(32) ‖ value(LE128) ‖ gasLimit(LE64) ‖ gasPrice(LE64)
//	         ‖ nonce(LE64) ‖ chainId(LE64) ‖ data )
func HashTransaction(tx Transaction) ([]byte, error) {
	fromBytes, err := FromOmneAddress(tx.From)
	if err != nil {
		return nil, fmt.Errorf("from: %w", err)
	}
	var toBytes []byte
	if tx.To != "" {
		if toBytes, err = FromOmneAddress(tx.To); err != nil {
			return nil, fmt.Errorf("to: %w", err)
		}
	}
	var dataBytes []byte
	if tx.Data != "" {
		if dataBytes, err = FromHex(tx.Data); err != nil {
			return nil, fmt.Errorf("data: %w", err)
		}
	}
	gasPrice, ok := new(big.Int).SetString(tx.GasPrice, 10)
	if !ok {
		return nil, fmt.Errorf("invalid gasPrice: %q", tx.GasPrice)
	}

	buf := make([]byte, 0, 32+32+16+8+8+8+8+len(dataBytes))
	buf = append(buf, fromBytes...)
	buf = append(buf, toBytes...)
	buf = append(buf, leBig(tx.Value, 16)...)
	buf = append(buf, leUint(tx.GasLimit, 8)...)
	buf = append(buf, leUint(gasPrice.Uint64(), 8)...)
	buf = append(buf, leUint(tx.Nonce, 8)...)
	buf = append(buf, leUint(uint64(tx.ChainID), 8)...)
	buf = append(buf, dataBytes...)

	h := sha256.Sum256(buf)
	return h[:], nil
}

File: python/src/omne_sdk/transaction.py
"""Transaction building + canonical hashing.

The signing preimage matches the Rust-side hash_transaction() in
omne-blockchain/src/rpc/wallet.rs and the TS wallet: fields are concatenated in
a fixed order with little-endian numeric encoding, then SHA-256'd. The node
reconstructs this exact preimage from the wire payload to verify the signature,
so every signed field must round-trip byte-for-byte.

    SHA-256( from(32) ‖ to(32) ‖ value(LE128) ‖ gasLimit(LE64) ‖ gasPrice(LE64)
             ‖ nonce(LE64) ‖ chainId(LE64) ‖ data )

Addresses are decoded from their om1z form to 32 raw bytes for the preimage.
`data` is the hex-encoded ABI calldata (see abi.encode_contract_call).
"""

from __future__ import annotations

import hashlib

from .address import from_omne_address, from_hex

DEFAULT_GAS_LIMIT = 200_000
DEFAULT_GAS_PRICE = "5000"


def _le(value: int, num_bytes: int) -> bytes:
    return int(value).to_bytes(num_bytes, "little")


def hash_transaction(transaction: dict) -> bytes:
    """Canonical 32-byte transaction hash used as the ML-DSA-44 signing message."""
    if transaction.get("chainId") is None:
        raise ValueError("chainId must be set on the transaction before hashing")

    from_bytes = from_omne_address(transaction["from"])
    to = transaction.get("to")
    to_bytes = from_omne_address(to) if to else b""

    data = transaction.get("data") or ""
    data_bytes = from_hex(data) if data else b""

    preimage = b"".join(
        [
            from_bytes,
            to_bytes,
            _le(int(transaction["value"]), 16),
            _le(int(transaction["gasLimit"]), 8),
            _le(int(transaction["gasPrice"]), 8),
            _le(int(transaction["nonce"]), 8),
            _le(int(transaction["chainId"]), 8),
            data_bytes,
        ]
    )
    return hashlib.sha256(preimage).digest()


def build_transaction(
    *,
    sender: str,
    to: str,
    data: str = "",
    value: int | str = 0,
    gas_limit: int = DEFAULT_GAS_LIMIT,
    gas_price: str = DEFAULT_GAS_PRICE,
    nonce: int = 0,
    chain_id: int | None = None,
) -> dict:
    """Build an unsigned transaction dict ready for WalletAccount.sign_transaction."""
    tx: dict = {
        "from": sender,
        "to": to,
        "value": str(value),
        "gasLimit": gas_limit,
        "gasPrice": gas_price,
        "nonce": nonce,
        "data": data,
    }
    if chain_id is not None:
        tx["chainId"] = chain_id
    return tx
